electronics_lib/PowerConditioning.py

```diff
@@ -263,3 +264,136 @@ def connected_from(self, gnd: Optional[Port[VoltageLink]] = None, pwr_hi: Option
     if pwr_lo is not None:
       cast(Block, builder.get_enclosing_block()).connect(pwr_lo, self.pwr_lo)
     return self
+
+
+class PmosReverseProtection(PowerConditioner, KiCadSchematicBlock, Block):
+  """
+  - This circuit uses a PMOS as a switch, which turns on when the power supply is correctly connected,
+  allowing current to flow to the load. If the power supply polarity is reversed, the PMOS turns off,
+  disconnecting the load and protecting the circuit.
+  - This method is preferred over diode-based protection due to lower power loss
+  - In most cases, 100R-330R is good if there are chances for the appearance of sudden reverse voltage in the circuit.
+  - But if there are no chances of sudden reverse voltage during the continuous working of the circuit, anything from the 1k-50k resistor value can be used.
+  - Ref: https://components101.com/articles/design-guide-pmos-mosfet-for-reverse-voltage-polarity-protection
+  """
+  @init_in_parent
+  def __init__(self, gate_resistor: RangeLike = 10 * kOhm(tol=0.05),):
+    super().__init__()
+    self.gnd = self.Port(Ground.empty(), [Common])
+    self.pwr_in = self.Port(VoltageSink.empty())  # high-priority higher-voltage source
+    self.pwr_out = self.Port(VoltageSource.empty())
+
+    self.gate_resistor = self.ArgParameter(gate_resistor)
+
+  def contents(self):
+    super().contents()
+
+    output_current_draw = self.pwr_out.link().current_drawn
+    # Pfet
+    self.fet = self.Block(Fet.PFet(
+      drain_voltage=(0, self.pwr_out.link().voltage.upper()),
+      drain_current=output_current_draw,
+      # gate voltage accounts for a possible power on transient
+      gate_voltage=(- self.pwr_out.link().voltage.upper(),  self.pwr_out.link().voltage.upper()),
+    ))
+
+    self.res = self.Block(Resistor(self.gate_resistor))
+    # TODO: generate zener diode for high voltage applications
+    #  self.diode = self.Block(ZenerDiode(self.clamp_voltage))
+
+    self.import_kicad(
+      self.file_path("resources", f"{self.__class__.__name__}.kicad_sch"),
+      conversions={
+        'pwr_in': VoltageSink(
+          current_draw=output_current_draw,
+        ),
+        'pwr_out': VoltageSource(
+          voltage_out=self.pwr_in.link().voltage,
+         ),
+        'gnd': Ground(),
+      })
+
+
+
+class PmosChargerReverseProtection(PowerConditioner, KiCadSchematicBlock, Block):
+  class PmosChargerReverseProtection(PowerConditioner, KiCadSchematicBlock, Block):
+    """PMOS Charger Reverse Protection.
```

Comment:
Similar comment to the first sentence in the PmosReverseProtection docstring, except this also supports charging.